test: fixup PySpark test failures

We currently get the following PySpark test failures:
```python
FAILED tests/expr_and_series/str/to_uppercase_to_lowercase_test.py::test_str_to_uppercase[pyspark-data1-expected1] - AssertionError: Mismatch at index 0: SPECIAL CASE SS != SPECIAL CASE ẞ
Expected: {'a': ['SPECIAL CASE ẞ', 'ΣPECIAL CAΣE']}
Got: {'a': ['SPECIAL CASE SS', 'ΣPECIAL CAΣE']}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate_custom[pyspark-7m-expected4] - ValueError: Only multiple 1 is currently supported for Spark-like.
Got 7.
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1ms-expected2] - AssertionError: Mismatch at index 0: 2021-03-01 12:34:56.049000+00:00 != 2021-03-01 12:34:56.049000
Expected: {'a': [datetime.datetime(2021, 3, 1, 12, 34, 56, 49000), datetime.datetime(2020, 1, 2, 2, 4, 14, 715000)]}
Got: {'a': [datetime.datetime(2021, 3, 1, 12, 34, 56, 49000, tzinfo=<UTC>), datetime.datetime(2020, 1, 2, 2, 4, 14, 715000, tzinfo=<UTC>)]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate_custom[pyspark-2q-expected8] - ValueError: Only multiple 1 is currently supported for Spark-like.
Got 2.
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1mo-expected7] - AssertionError: Mismatch at index 0: 2021-03-01 00:00:00+00:00 != 2021-03-01 00:00:00
Expected: {'a': [datetime.datetime(2021, 3, 1, 0, 0), datetime.datetime(2020, 1, 1, 0, 0)]}
Got: {'a': [datetime.datetime(2021, 3, 1, 0, 0, tzinfo=<UTC>), datetime.datetime(2020, 1, 1, 0, 0, tzinfo=<UTC>)]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1h-expected5] - AssertionError: Mismatch at index 0: 2021-03-01 12:00:00+00:00 != 2021-03-01 12:00:00
Expected: {'a': [datetime.datetime(2021, 3, 1, 12, 0), datetime.datetime(2020, 1, 2, 2, 0)]}
Got: {'a': [datetime.datetime(2021, 3, 1, 12, 0, tzinfo=<UTC>), datetime.datetime(2020, 1, 2, 2, 0, tzinfo=<UTC>)]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate_custom[pyspark-13d-expected6] - ValueError: Only multiple 1 is currently supported for Spark-like.
Got 13.
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate_custom[pyspark-2ns-expected0] - ValueError: Only multiple 1 is currently supported for Spark-like.
Got 2.
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1m-expected4] - AssertionError: Mismatch at index 0: 2021-03-01 12:34:00+00:00 != 2021-03-01 12:34:00
Expected: {'a': [datetime.datetime(2021, 3, 1, 12, 34), datetime.datetime(2020, 1, 2, 2, 4)]}
Got: {'a': [datetime.datetime(2021, 3, 1, 12, 34, tzinfo=<UTC>), datetime.datetime(2020, 1, 2, 2, 4, tzinfo=<UTC>)]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate_custom[pyspark-2us-expected1] - ValueError: Only multiple 1 is currently supported for Spark-like.
Got 2.
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1ns-expected0] - AssertionError: Mismatch at index 0: None != 2021-03-01 12:34:56.049012
Expected: {'a': [datetime.datetime(2021, 3, 1, 12, 34, 56, 49012), datetime.datetime(2020, 1, 2, 2, 4, 14, 715123)]}
Got: {'a': [None, None]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1d-expected6] - AssertionError: Mismatch at index 0: 2021-03-01 00:00:00+00:00 != 2021-03-01 00:00:00
Expected: {'a': [datetime.datetime(2021, 3, 1, 0, 0), datetime.datetime(2020, 1, 2, 0, 0)]}
Got: {'a': [datetime.datetime(2021, 3, 1, 0, 0, tzinfo=<UTC>), datetime.datetime(2020, 1, 2, 0, 0, tzinfo=<UTC>)]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate_custom[pyspark-10s-expected3] - ValueError: Only multiple 1 is currently supported for Spark-like.
Got 10.
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1us-expected1] - AssertionError: Mismatch at index 0: 2021-03-01 12:34:56.049012+00:00 != 2021-03-01 12:34:56.049012
Expected: {'a': [datetime.datetime(2021, 3, 1, 12, 34, 56, 49012), datetime.datetime(2020, 1, 2, 2, 4, 14, 715123)]}
Got: {'a': [datetime.datetime(2021, 3, 1, 12, 34, 56, 49012, tzinfo=<UTC>), datetime.datetime(2020, 1, 2, 2, 4, 14, 715123, tzinfo=<UTC>)]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1y-expected9] - AssertionError: Mismatch at index 0: 2021-01-01 00:00:00+00:00 != 2021-01-01 00:00:00
Expected: {'a': [datetime.datetime(2021, 1, 1, 0, 0), datetime.datetime(2020, 1, 1, 0, 0)]}
Got: {'a': [datetime.datetime(2021, 1, 1, 0, 0, tzinfo=<UTC>), datetime.datetime(2020, 1, 1, 0, 0, tzinfo=<UTC>)]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1s-expected3] - AssertionError: Mismatch at index 0: 2021-03-01 12:34:56+00:00 != 2021-03-01 12:34:56
Expected: {'a': [datetime.datetime(2021, 3, 1, 12, 34, 56), datetime.datetime(2020, 1, 2, 2, 4, 14)]}
Got: {'a': [datetime.datetime(2021, 3, 1, 12, 34, 56, tzinfo=<UTC>), datetime.datetime(2020, 1, 2, 2, 4, 14, tzinfo=<UTC>)]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate_custom[pyspark-7h-expected5] - ValueError: Only multiple 1 is currently supported for Spark-like.
Got 7.
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate[pyspark-1q-expected8] - AssertionError: Mismatch at index 0: 2021-01-01 00:00:00+00:00 != 2021-01-01 00:00:00
Expected: {'a': [datetime.datetime(2021, 1, 1, 0, 0), datetime.datetime(2020, 1, 1, 0, 0)]}
Got: {'a': [datetime.datetime(2021, 1, 1, 0, 0, tzinfo=<UTC>), datetime.datetime(2020, 1, 1, 0, 0, tzinfo=<UTC>)]}
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate_custom[pyspark-2ms-expected2] - ValueError: Only multiple 1 is currently supported for Spark-like.
Got 2.
FAILED tests/expr_and_series/dt/truncate_test.py::test_truncate_custom[pyspark-3mo-expected7] - ValueError: Only multiple 1 is currently supported for Spark-like.
Got 3.
```

I don't think any of them are blockers, but we should address them

--- a/narwhals/_spark_like/expr_dt.py
+++ b/narwhals/_spark_like/expr_dt.py
@@ -71,6 +71,9 @@ class SparkLikeExprDateTimeNamespace:
         if multiple != 1:
             msg = f"Only multiple 1 is currently supported for Spark-like.\nGot {multiple!s}."
             raise ValueError(msg)
+        if unit == "ns":
+            msg = "Truncating to nanoseconds is not yet supported for Spark-like."
+            raise NotImplementedError(msg)
         format = UNITS_DICT[unit]
 
         def _truncate(_input: Column) -> Column:

--- a/tests/expr_and_series/dt/truncate_test.py
+++ b/tests/expr_and_series/dt/truncate_test.py
@@ -99,23 +99,19 @@ def test_truncate(
     every: str,
     expected: list[datetime],
 ) -> None:
-    if any(x in str(constructor) for x in ("sqlframe",)):
+    if any(x in str(constructor) for x in ("sqlframe", "pyspark")):
+        # TODO(marco): investigate pyspark, it also localizes to UTC here.
         request.applymarker(
             pytest.mark.xfail(reason="https://github.com/eakmanrq/sqlframe/issues/383")
         )
-    if every.endswith("ns") and any(x in str(constructor) for x in ("polars", "duckdb")):
+    if every.endswith("ns") and any(
+        x in str(constructor) for x in ("polars", "duckdb", "pyspark")
+    ):
         request.applymarker(pytest.mark.xfail())
     if any(every.endswith(x) for x in ("mo", "q", "y")) and any(
         x in str(constructor) for x in ("dask", "cudf")
     ):
         request.applymarker(pytest.mark.xfail(reason="Not implemented"))
-    request.applymarker(
-        pytest.mark.xfail(
-            ("pyspark" in str(constructor) and every.endswith("ns")),
-            raises=NotImplementedError,
-            reason="Truncating to nanoseconds is not yet supported",
-        )
-    )
     df = nw.from_native(constructor(data))
     result = df.select(nw.col("a").dt.truncate(every))
     assert_equal_data(result, {"a": expected})

--- a/tests/expr_and_series/dt/truncate_test.py
+++ b/tests/expr_and_series/dt/truncate_test.py
@@ -182,7 +182,7 @@ def test_truncate(
         ),
     ],
 )
-def test_truncate_custom(
+def test_truncate_multiples(
     request: pytest.FixtureRequest,
     constructor: Constructor,
     every: str,
@@ -201,6 +201,13 @@ def test_truncate_custom(
         x in str(constructor) for x in ("dask",)
     ):
         request.applymarker(pytest.mark.xfail(reason="Not implemented"))
+    request.applymarker(
+        pytest.mark.xfail(
+            "pyspark" in str(constructor),
+            raises=ValueError,
+            reason="Only multiple 1 is currently supported",
+        )
+    )
     df = nw.from_native(constructor(data))
     result = df.select(nw.col("a").dt.truncate(every))
     assert_equal_data(result, {"a": expected})

--- a/tests/expr_and_series/dt/truncate_test.py
+++ b/tests/expr_and_series/dt/truncate_test.py
@@ -109,6 +109,13 @@ def test_truncate(
         x in str(constructor) for x in ("dask", "cudf")
     ):
         request.applymarker(pytest.mark.xfail(reason="Not implemented"))
+    request.applymarker(
+        pytest.mark.xfail(
+            ("pyspark" in str(constructor) and every.endswith("ns")),
+            raises=NotImplementedError,
+            reason="Truncating to nanoseconds is not yet supported",
+        )
+    )
     df = nw.from_native(constructor(data))
     result = df.select(nw.col("a").dt.truncate(every))
     assert_equal_data(result, {"a": expected})

--- a/tests/expr_and_series/str/to_uppercase_to_lowercase_test.py
+++ b/tests/expr_and_series/str/to_uppercase_to_lowercase_test.py
@@ -35,9 +35,16 @@ def test_str_to_uppercase(
     if (
         any(
             x in str(constructor)
-            for x in ("pandas_constructor", "pandas_nullable", "polars", "cudf")
+            for x in (
+                "pandas_constructor",
+                "pandas_nullable",
+                "polars",
+                "cudf",
+                "pyspark",
+            )
         )
         and "ẞ" in expected["a"][0]
+        and "sqlframe" not in str(constructor)
     ):
         expected = {"a": ["SPECIAL CASE SS", "ΣPECIAL CAΣE"]}
 
@@ -72,9 +79,16 @@ def test_str_to_uppercase_series(
     if (
         any(
             x in str(constructor_eager)
-            for x in ("pandas_constructor", "pandas_nullable", "polars", "cudf")
+            for x in (
+                "pandas_constructor",
+                "pandas_nullable",
+                "polars",
+                "cudf",
+                "pyspark",
+            )
         )
         and "ẞ" in expected["a"][0]
+        and "sqlframe" not in str(constructor_eager)
     ):
         expected = {"a": ["SPECIAL CASE SS", "ΣPECIAL CAΣE"]}
 

